extensions/src/AWSSDK.Extensions.CborProtocol/Internal/Transform/CborUnmarshallerContext.cs

```diff
@@ -0,0 +1,213 @@
+                    && requestContext?.OriginalRequest?.CoreChecksumMode
+                        == CoreChecksumResponseBehavior.ENABLED
+                )
+                {
+                    base.SetupCRCStream(responseData, responseStream, contentLength);
+                    base.SetupFlexibleChecksumStream(
+                        responseData,
+                        CrcStream ?? responseStream,
+                        contentLength,
+                        requestContext
+                    );
+                }
+                if (parsedContentLengthHeader && responseData.ContentLength == contentLength)
+                {
+                    streamSize = contentLength;
+                }
+            }
+
+            this.WebResponseData = responseData;
+            this.MaintainResponseBody = maintainResponseBody;
+            this.IsException = isException;
+
+            Reader = CreateCborReaderFromStream(
+                FlexibleChecksumStream ?? CrcStream ?? responseStream,
+                streamSize
+            );
+        }
+
+        private CborReader CreateCborReaderFromStream(Stream stream, long? streamSize = null)
+        {
+            int totalRead = 0;
+            if (streamSize.HasValue)
+            {
+                // If we know the size, we can read directly into a buffer of exact size
+                var buffer = new byte[streamSize.Value];
+
+                while (totalRead < streamSize.Value)
+                {
+                    int bytesRead = stream.Read(buffer, totalRead, (int)(streamSize.Value - totalRead));
+                    if (bytesRead == 0)
+                    {
+                        // If no bytes are read, it means we've reached the end of the stream before reading the whole streamSize.
+                        throw new EndOfStreamException($"Expected {streamSize.Value} bytes but only read {totalRead}.");
+                    }
+
+                    totalRead += bytesRead;
+                }
+
+                return new CborReader(buffer);
+            }
+
+            const int InitialBufferSize = 1024 * 8; // 8kb
+            rentedBuffer = ArrayPool<byte>.Shared.Rent(InitialBufferSize);
+
+            while (true)
+            {
+                int read = stream.Read(rentedBuffer, totalRead, rentedBuffer.Length - totalRead);
+                if (read == 0)
+                    break;
+
+                totalRead += read;
+
+                if (totalRead == rentedBuffer.Length)
+                {
+                    // Expand the buffer size by doubling it
+                    var newBuffer = ArrayPool<byte>.Shared.Rent(rentedBuffer.Length * 2);
+                    try
+                    {
+                        Buffer.BlockCopy(rentedBuffer, 0, newBuffer, 0, totalRead);
+                    }
+                    catch
+                    {
+                        ArrayPool<byte>.Shared.Return(newBuffer);
```

Comment:
Since we are exiting out due to an exception shouldn't we also `Return` the `rentedBuffer` as well?